core/ir/instr_shared.c

```diff
@@ -176,6 +176,16 @@ instr_init(void *drcontext, instr_t *instr)
      * an uninitialized instruction */
     memset((void *)instr, 0, sizeof(instr_t));
     instr_set_isa_mode(instr, dr_get_isa_mode(dcontext));
+#ifndef NOT_DYNAMORIO_CORE_PROPER
+    /* Just like in global_heap_alloc() we pay the cost of this check to support
+     * drdecode use even with full DR linked in (i#2499).  Decoding of simple
+     * single-source-no-dest instrs never hits the heap code so we check here too.
+     */
+    if (dcontext == GLOBAL_DCONTEXT && !dynamo_heap_initialized) {
+        /* XXX: We have no control point to call standalone_exit(). */
+        standalone_init();
```

Comment:
I have mixed feelings: calling dr_standalone_init() is not supposed to be necessary (and in fact is it even possible if you only link drdecode?  Does drdecode even export that?  I don't think it does: so if the other tools like opcode_mix all stopped pulling in the full libdynamorio.so that would not be a feasible solution) as you're supposed to be able to use drdecode as a supposedly very simple library with no init required, and we already have the heap.c call which is supposed to be good enough.

It is possible someone might add some init-time setup that requires teardown for drdecode: maybe we could assume libc and register atexit or something.  But given that we already have the heap.c I'm thinking we file a separate issue on that.  I also thought there was an issue on downgrading our tools and other drcachesim/ dir uses of libDR down to drdecode: if I can't find it we should file a new one, since these tools just doing decoding do not need full DR.  Though I think raw2trace for module map is using dr_map_file: so maybe for legacy traces we do still need full DR.

Reply:
Let's just use this simpler workaround for now.

> It is possible someone might add some init-time setup that requires teardown for drdecode

My only concern at the moment was this. But I guess a comment should do fine for now.

Reply:
Per offline discussion: sticking with the current not-ideal-but-simple implicit-init scheme.  I re-opened #2499 to cover trying to make that scheme cleaner.  I added a `TODO i#2499` here and in the heap.c trigger.